relay.go

```diff
@@ -17,8 +24,85 @@ type RelayHosts interface {
 }
 
 type relayItem struct {
+	*time.Timer
+
 	remapID     uint32
 	destination *Relayer
+	tomb        bool
+}
+
+type relayItems struct {
+	sync.RWMutex
+
+	logger Logger
+	tombs  uint64
+	items  map[uint32]relayItem
+}
+
+func newRelayItems(logger Logger) *relayItems {
+	return &relayItems{
+		items: make(map[uint32]relayItem),
+	}
+}
+
+// Count returns the number of items in the relay.
+func (r *relayItems) Count() int {
+	r.RLock()
+	n := len(r.items)
+	r.RUnlock()
+	return n
+}
+
+// Get checks for a relay item by ID. It returns the item, if it's still active,
+// whether the item is a tombstone,
+func (r *relayItems) Get(id uint32) (relayItem, bool) {
+	r.RLock()
+	item, ok := r.items[id]
+	r.RUnlock()
+
+	return item, ok
+}
+
+// Add adds a relay item.
+func (r *relayItems) Add(id uint32, item relayItem) {
+	r.Lock()
+	r.items[id] = item
+	r.Unlock()
+}
+
+// Delete removes a relayItem completely (without leaving a tombstone).
+func (r *relayItems) Delete(id uint32) {
+	r.Lock()
+	if item, ok := r.items[id]; ok {
+		item.Stop()
+		if item.tomb {
+			r.tombs--
+		}
+	}
+	delete(r.items, id)
+	r.Unlock()
+}
+
+// Entomb sets the tomb bit on a relayItem and schedules a garbage collection.
+func (r *relayItems) Entomb(id uint32, deleteAfter time.Duration) {
+	r.Lock()
+	if r.tombs > _maxRelayTombs {
+		r.logger.WithFields(LogField{"id", id}).Warn("Too many tombstones, deleting relay item immediately.")
+		r.Unlock()
+		return
+	}
+	item, ok := r.items[id]
+	if !ok {
+		r.logger.WithFields(LogField{"id", id}).Warn("Can't find relay item to entomb.")
+		r.Unlock()
+		return
+	}
+	r.tombs++
+	item.tomb = true
+	r.items[id] = item
+	r.Unlock()
+
+	time.AfterFunc(deleteAfter, func() { r.Delete(id) })
```

Comment:
(no action needed), long term, do you think we should clear these out in batches?